packages/shared/src/hooks/useCompanionTrigger.tsx

```diff
@@ -0,0 +1,167 @@
+import React, { useCallback, useMemo } from 'react';
+import { useGrowthBook } from '@growthbook/growthbook-react';
+import { FeedPostClick } from './feed/useFeedOnPostClick';
+import { Post } from '../graphql/posts';
+import { Features } from '../lib/featureManagement';
+import { useAnalyticsContext } from '../contexts/AnalyticsContext';
+
+import { useExtensionPermission } from './useExtensionPermission';
+import { useLazyModal } from './useLazyModal';
+import { LazyModal } from '../components/modals/common/types';
+import { useActions } from './useActions';
+import { ActionType } from '../graphql/actions';
+import { AnalyticsEvent, Origin } from '../lib/analytics';
+
+type CompanionTriggerProps = {
+  post: Post;
+  index?: number;
+  row?: number;
+  column?: number;
+};
+
+type OpenArticle = (data?: CompanionTriggerProps) => () => Promise<void>;
+
+export type CompanionTrigger = {
+  onFeedArticleClick: FeedPostClick;
+  onPostArticleClick: PostClick;
+};
+
+type PostClick = (e: React.MouseEvent, data: { post: Post }) => Promise<void>;
+
+export default function useCompanionTrigger(
+  customPostClickHandler: (
+    post: Post,
+    index?: number,
+    row?: number,
+    column?: number,
+  ) => Promise<void>,
+): CompanionTrigger {
+  const isExtension = process.env.TARGET_BROWSER;
+  const { trackEvent } = useAnalyticsContext();
+  const { closeModal: closeLazyModal, openModal: openLazyModal } =
+    useLazyModal();
+  const { isActionsFetched, checkHasCompleted, completeAction } = useActions();
+  const gb = useGrowthBook();
+
+  const featureEnabled = useMemo(() => {
+    // gb.isOn triggers enrollment in the experiment
+    // to avoid noise in the data, only call this if we are in the extension
+    if (isExtension)
+      return gb.isOn(Features.EngagementLoopJuly2023Companion.id);
+
+    return false;
+  }, [gb, isExtension]);
+
+  const { requestContentScripts, useContentScriptStatus } =
+    useExtensionPermission({
+      origin: 'companion modal permission button',
+    });
+
+  const { contentScriptGranted } = useContentScriptStatus();
+  const alreadyCompleted = checkHasCompleted(
+    ActionType.EngagementLoopJuly2023CompanionModal,
+  );
+
+  const handleCompleteAction = useCallback(() => {
+    if (isActionsFetched)
+      completeAction(ActionType.EngagementLoopJuly2023CompanionModal);
```

Comment:
You don't have to wait for this because we merge any locally completed actions once they are fetch so you can just call `completeAction`

Reply:
👍  thank you for pointing that out. But if it does not matter, isn't it better to be more explicit?

Reply:
When I think about it, with condition like this there could be a case where actions fetch request does not resolve and user closes the modal which would mean action will not be completed and user will see the modal once again, which is not per spec. 

To your point thought `isActionsFetched` is more important when reading since you probably want to wait server state before deciding action is completed. So if that was the case we would use it 👍

Reply:
I think I tested this and if `isActionFetched` is `undefined`, `completeAction` fails as well

Reply:
Please double check 🙏 if it does not work thats a bug.

Reply:
Looks like this works fine, I've updated and cleaned it up a little 👍